File: scraper/self_sufficiency.py
"""Self-Sufficiency Score — Python port of frontend/src/utils/selfSufficiency.ts.

Pre-computes the headline autonomy metric at scrape time and stamps
`selfSufficiency: { composite, axes }` onto each listing in the slim
index. Lets the Browse page render axis bars + filter on per-axis
minimums without shipping `geoEnrichment` through the index — the
biggest single contributor to listings.json's bulk.

The TS version (used on the detail page) recomputes the full report
including per-axis gap interventions and prose verdicts. Those are
detail-only and stay client-side since the detail page fetches the
full per-id record. The Python port emits just the numerical core:

    {
        "composite": 67,
        "axes": [
            {"key": "food", "score": 22, "weight": 0.28},
            {"key": "water", "score": 98, "weight": 0.25},
            ...
        ]
    }

Keep the constants / band thresholds in sync with the TS version. If
you tweak one, tweak the other — they're calibrated together.
"""

from __future__ import annotations

from typing import Any

AXIS_WEIGHTS = {
    "food": 0.28,
    "water": 0.25,
    "energy": 0.20,
    "shelter": 0.15,
    "resilience": 0.12,
}

CAP_SCORES = {1: 100, 2: 88, 3: 75, 4: 60, 5: 45, 6: 35, 7: 22, 8: 10}


def _clip(v: float) -> int:
    return max(0, min(100, round(v)))


def _features(p: dict[str, Any]) -> set[str]:
    return set(p.get("features") or [])


def _improvements(p: dict[str, Any]) -> dict[str, Any]:
    return p.get("improvements") or {}


def _food(p: dict[str, Any]) -> int:
    soil = (p.get("geoEnrichment") or {}).get("soil") or {}
    flood = (p.get("geoEnrichment") or {}).get("flood") or {}
    acres = float(p.get("acreage") or 0)
    features = _features(p)

    try:
        cap = int(soil.get("capabilityClass") or 0) or 5
    except (TypeError, ValueError):
        cap = 5
    cap_score = CAP_SCORES.get(cap, 45)

    # Acreage gate — same shape as TS: min(1, max(0.3, acres/5))
    acre_factor = min(1.0, max(0.3, acres / 5.0))
    score = cap_score * acre_factor

    drain = (soil.get("drainageClass") or "").lower()
    if "poorly" in drain:
        score -= 10
    if flood.get("isSFHA"):
        score -= 12
    if "pasture" in features:
        score += 6
    if "timber" in features:
        score += 4
    return _clip(score)


def _water(p: dict[str, Any]) -> int:
    features = _features(p)
    improvements = _improvements(p)
    proximity = (p.get("geoEnrichment") or {}).get("proximity") or {}
    flood = (p.get("geoEnrichment") or {}).get("flood") or {}
    location = p.get("location") or {}
    lat = float(location.get("lat") or 0)
    lng_abs = abs(float(location.get("lng") or 0))

    score = 30
    has_well = "water_well" in features or improvements.get("well")
    has_city = improvements.get("water_city")
    if has_well or has_city:
        score += 35

    named_water = [str(n).lower() for n in (proximity.get("namedWaterFeatures") or [])]
    has_flowing = "water_creek" in features or any(
        any(k in n for k in ("creek", "stream", "run", "river")) for n in named_water
    )
    if has_flowing:
        score += 18
    if "water_pond" in features:
        score += 10

    # Desert SW penalty: lat 30-38, lng 103-114 (interior arid US)
    desert_sw = 30 < lat < 38 and 103 < lng_abs < 114
    if not desert_sw:
        score += 5

    if flood.get("isSFHA"):
        score -= 8
    return _clip(score)


def _energy(p: dict[str, Any]) -> int:
    location = p.get("location") or {}
    lat = abs(float(location.get("lat") or 0))
    acres = float(p.get("acreage") or 0)
    features = _features(p)
    improvements = _improvements(p)
    soil = (p.get("geoEnrichment") or {}).get("soil") or {}

    # Solar baseline by latitude (matches the TS piecewise table)
    if lat < 32:
        sun_hours = 6.0
    elif lat < 36:
        sun_hours = 5.5
    elif lat < 40:
        sun_hours = 5.0
    elif lat < 44:
        sun_hours = 4.5
    else:
        sun_hours = 4.2
    score = (sun_hours / 6.0) * 60.0

    if "timber" in features or (acres >= 10 and "hunting" in features):
        score += 12
    if acres >= 1.5:
        score += 6
    try:
        slope = float(soil.get("slopePercent") or 0)
    except (TypeError, ValueError):
        slope = 0.0
    if "water_creek" in features and slope >= 5:
        score += 10
    if improvements.get("electric") or "electric" in features:
        score += 8
    return _clip(score)


def _shelter(p: dict[str, Any]) -> int:
    features = _features(p)
    improvements = _improvements(p)
    acres = float(p.get("acreage") or 0)

    score = 20
    if improvements.get("home") or improvements.get("cabin"):
        score += 50
    elif improvements.get("outbuilding") or improvements.get("barn"):
        score += 20
    if "structures" in features:
        score += 10
    if "timber" in features and acres >= 5:
        score += 15
    if improvements.get("well") or "water_well" in features:
        score += 8
    if improvements.get("septic") or "septic" in features:
        score += 8
    if improvements.get("electric") or "electric" in features:
        score += 6
    if p.get("moveInReady"):
        score = max(score, 90)
    return _clip(score)


def _resilience(p: dict[str, Any]) -> int:
    features = _features(p)
    flood = (p.get("geoEnrichment") or {}).get("flood") or {}
    soil = (p.get("geoEnrichment") or {}).get("soil") or {}
    location = p.get("location") or {}
    lat = float(location.get("lat") or 0)
    lng_abs = abs(float(location.get("lng") or 0))
    acres = float(p.get("acreage") or 0)

    score = 75
    if flood.get("isSFHA"):
        score -= 30
    zone = flood.get("floodZone")
    if zone in ("X", "X500"):
        score += 5

    # Arid SW interior — drought proxy (lat 30-38, lng 103-116)
    if 30 < lat < 38 and 103 < lng_abs < 116:
        score -= 12

    try:
        slope = float(soil.get("slopePercent") or 0)
    except (TypeError, ValueError):
        slope = 0.0
    if slope > 20:
        score -= 10
    elif slope > 12:
        score -= 5

    if acres >= 10:
        score += 8
    if "no_hoa" in features:
        score += 5
    return _clip(score)


def compute_self_sufficiency(p: dict[str, Any]) -> dict[str, Any]:
    """Return the slim {composite, axes} payload for a listing dict.

    `axes` is a list of {key, score, weight} entries in the canonical
    order Food → Water → Energy → Shelter → Resilience. Composite is
    the weighted average, rounded to int.
    """
    food = _food(p)
    water = _water(p)
    energy = _energy(p)
    shelter = _shelter(p)
    resilience = _resilience(p)
    composite = round(
        food * AXIS_WEIGHTS["food"]
        + water * AXIS_WEIGHTS["water"]
        + energy * AXIS_WEIGHTS["energy"]
        + shelter * AXIS_WEIGHTS["shelter"]
        + resilience * AXIS_WEIGHTS["resilience"]
    )
    return {
        "composite": composite,
        "axes": [
            {"key": "food", "score": food, "weight": AXIS_WEIGHTS["food"]},
            {"key": "water", "score": water, "weight": AXIS_WEIGHTS["water"]},
            {"key": "energy", "score": energy, "weight": AXIS_WEIGHTS["energy"]},
            {"key": "shelter", "score": shelter, "weight": AXIS_WEIGHTS["shelter"]},
            {
                "key": "resilience",
                "score": resilience,
                "weight": AXIS_WEIGHTS["resilience"],
            },
        ],
    }

File: frontend/src/utils/selfSufficiency.ts
import { Property } from '../types/property';

/**
 * Self-Sufficiency Score — the headline metric for the redesigned
 * detail page. Replaces Deal / Investment / Homestead-Fit as the
 * lead score. Asks one question: how close is this parcel to fully
 * autonomous living?
 *
 * Five weighted axes; each independently scored 0–100. Composite
 * is a weighted average. Heuristic — calibrated against the data
 * we already have stamped on each listing (geoEnrichment, features,
 * improvements, location, acreage). No new data fetches.
 */

export type AxisKey = 'food' | 'water' | 'energy' | 'shelter' | 'resilience';

export interface Axis {
  key: AxisKey;
  label: string;
  score: number;
  weight: number;
  /** One-line plain-language verdict for this axis. */
  verdict: string;
  /** Specific interventions that would lift this axis, with rough
   *  cost and score-delta. Drives the buildout-to-autonomy panel. */
  gaps: Gap[];
}

export interface Gap {
  label: string;
  costLowUsd: number;
  costHighUsd: number;
  /** Estimated score lift on this axis if executed. */
  liftPoints: number;
}

export interface SelfSufficiencyReport {
  composite: number;
  axes: Axis[];
  weakest: Axis;
  /** Total cost range to bring every axis to ≥85. */
  costToFullLowUsd: number;
  costToFullHighUsd: number;
  /** Composite if all gaps closed. */
  potentialComposite: number;
}

const AXIS_WEIGHTS: Record<AxisKey, number> = {
  food: 0.28,
  water: 0.25,
  energy: 0.20,
  shelter: 0.15,
  resilience: 0.12,
};

// ── Per-axis calculators ───────────────────────────────────────────────

const foodAxis = (p: Property): Axis => {
  const soil = p.geoEnrichment?.soil;
  const acres = p.acreage ?? 0;
  const features = new Set(p.features ?? []);
  const cap = parseInt(soil?.capabilityClass ?? '', 10) || 5;
  const capScore = ({ 1: 100, 2: 88, 3: 75, 4: 60, 5: 45, 6: 35, 7: 22, 8: 10 } as Record<number, number>)[cap] ?? 45;

  // Acreage gate: <2ac caps the food score; >20ac doesn't keep raising it
  // because the bottleneck shifts to labor / capital, not land.
  const acreFactor = Math.min(1, Math.max(0.3, acres / 5));

  let score = capScore * acreFactor;
  const drain = (soil?.drainageClass ?? '').toLowerCase();
  if (drain.includes('poorly')) score -= 10;
  if (p.geoEnrichment?.flood?.isSFHA) score -= 12;
  if (features.has('pasture')) score += 6;
  if (features.has('timber')) score += 4; // mast crops, foraging

  score = Math.max(0, Math.min(100, Math.round(score)));

  const gaps: Gap[] = [];
  if (score < 70 && cap <= 4 && acres >= 1) {
    gaps.push({
      label: 'Build raised beds + greenhouse for intensive growing',
      costLowUsd: 4_000,
      costHighUsd: 15_000,
      liftPoints: 12,
    });
  }
  if (acres >= 5 && !features.has('pasture')) {
    gaps.push({
      label: 'Establish pasture / silvopasture (seeding + fencing)',
      costLowUsd: 6_000,
      costHighUsd: 18_000,
      liftPoints: 10,
    });
  }
  if (acres >= 10) {
    gaps.push({
      label: 'Plant orchard (50 fruit/nut trees, 5-yr to bearing)',
      costLowUsd: 1_500,
      costHighUsd: 4_500,
      liftPoints: 8,
    });
  }

  const verdict =
    score >= 80 ? `Excellent food potential — ${acres.toFixed(0)}ac of class-${cap} land.` :
    score >= 60 ? `Solid for diversified food production. Best-fit: pasture + orchard + market garden.` :
    score >= 40 ? `Workable with intensive methods (raised beds, high tunnels). Calorie autonomy harder.` :
    `Limited food capacity — better as woodlot or grazing than tillage.`;

  return { key: 'food', label: 'Food', score, weight: AXIS_WEIGHTS.food, verdict, gaps };
};

const waterAxis = (p: Property): Axis => {
  const features = new Set(p.features ?? []);
  const improvements = p.improvements ?? {};
  const proximity = p.geoEnrichment?.proximity;
  const namedWater = (proximity?.namedWaterFeatures ?? []).map((s) => s.toLowerCase());
  const lat = p.location?.lat ?? 0;

  let score = 30;
  let hasWell = features.has('water_well') || !!improvements.well;
  let hasCity = !!improvements.water_city;
  if (hasWell || hasCity) score += 35;
  if (features.has('water_creek') || namedWater.some((n) => /creek|stream|run|river/.test(n))) score += 18;
  if (features.has('water_pond')) score += 10;
  // Rough catchment potential: TX Hill Country / Austin metro gets ~30in/yr.
  // 30in × 1000sqft roof = ~18k gal/yr — meaningful supplemental supply.
  // Outside the desert SW (lat 30-38, lng -103 to -114), most US is workable.
  const lng = Math.abs(p.location?.lng ?? 0);
  const desertSW = lat > 30 && lat < 38 && lng > 103 && lng < 114;
  if (!desertSW) score += 5;

  if (p.geoEnrichment?.flood?.isSFHA) score -= 8; // ironic: too much water = bad

  score = Math.max(0, Math.min(100, Math.round(score)));

  const gaps: Gap[] = [];
  if (!hasWell && !hasCity) {
    gaps.push({
      label: 'Drill well (rural TX/MO/AR typical 200–500 ft)',
      costLowUsd: 8_000,
      costHighUsd: 25_000,
      liftPoints: 30,
    });
  }
  gaps.push({
    label: 'Add 10k gal rainwater cistern + first-flush + filtration',
    costLowUsd: 6_000,
    costHighUsd: 15_000,
    liftPoints: 12,
  });
  if (!features.has('water_pond') && (p.acreage ?? 0) >= 5) {
    gaps.push({
      label: 'Excavate ~1 acre × 8 ft pond (NRCS may cost-share)',
      costLowUsd: 5_000,
      costHighUsd: 20_000,
      liftPoints: 8,
    });
  }

  const verdict =
    score >= 80 ? 'Strong water security — well + surface water on parcel.' :
    score >= 60 ? 'Workable. Diversify with rainwater catchment for resilience.' :
    score >= 40 ? 'Tight. Drilling a well is the highest-leverage move.' :
    'Major bottleneck. No on-parcel source — bring water in or punt.';

  return { key: 'water', label: 'Water', score, weight: AXIS_WEIGHTS.water, verdict, gaps };
};

const energyAxis = (p: Property): Axis => {
  const lat = Math.abs(p.location?.lat ?? 0);
  const acres = p.acreage ?? 0;
  const features = new Set(p.features ?? []);
  const improvements = p.improvements ?? {};

  // Solar baseline by latitude.
  const sunHours = lat < 32 ? 6.0 : lat < 36 ? 5.5 : lat < 40 ? 5.0 : lat < 44 ? 4.5 : 4.2;
  let score = (sunHours / 6.0) * 60; // up to 60 from solar alone

  if (features.has('timber') || (acres >= 10 && features.has('hunting'))) score += 12; // wood/biomass
  if (acres >= 1.5) score += 6; // geothermal possible
  if (features.has('water_creek') && (p.geoEnrichment?.soil?.slopePercent ?? 0) >= 5) score += 10; // hydro
  if (improvements.electric || features.has('electric')) score += 8; // grid-tie option

  score = Math.max(0, Math.min(100, Math.round(score)));

  const gaps: Gap[] = [];
  gaps.push({
    label: 'Install 10 kW grid-tied solar + battery backup',
    costLowUsd: 22_000,
    costHighUsd: 45_000,
    liftPoints: 25,
  });
  if (acres >= 1.5) {
    gaps.push({
      label: 'Geothermal heat pump (3-ton, horizontal loop)',
      costLowUsd: 18_000,
      costHighUsd: 35_000,
      liftPoints: 12,
    });
  }
  if (features.has('water_creek') && (p.geoEnrichment?.soil?.slopePercent ?? 0) >= 5) {
    gaps.push({
      label: 'Micro-hydro (year-round flow, ≥25ft head)',
      costLowUsd: 8_000,
      costHighUsd: 25_000,
      liftPoints: 8,
    });
  }

  const verdict =
    score >= 80 ? 'Energy abundance — multiple paths to full autonomy.' :
    score >= 60 ? 'Workable. Solar + battery covers it; biomass adds resilience.' :
    score >= 40 ? 'Solar-only path. Real but uncomfortable in long winter overcasts.' :
    'Marginal. Northern latitude or shaded — supplement with biomass or grid-tie.';

  return { key: 'energy', label: 'Energy', score, weight: AXIS_WEIGHTS.energy, verdict, gaps };
};

const shelterAxis = (p: Property): Axis => {
  const features = new Set(p.features ?? []);
  const improvements = p.improvements ?? {};
  const acres = p.acreage ?? 0;

  let score = 20;
  if (improvements.home || improvements.cabin) score += 50;
  else if (improvements.outbuilding || improvements.barn) score += 20;
  if (features.has('structures')) score += 10;
  if (features.has('timber') && acres >= 5) score += 15; // can mill lumber
  if (improvements.well || features.has('water_well')) score += 8;
  if (improvements.septic || features.has('septic')) score += 8;
  if (improvements.electric || features.has('electric')) score += 6;
  if (p.moveInReady) score = Math.max(score, 90);

  score = Math.max(0, Math.min(100, Math.round(score)));

  const gaps: Gap[] = [];
  if (!improvements.home && !improvements.cabin) {
    gaps.push({
      label: 'Build cabin (modest 600sqft DIY → modular turnkey)',
      costLowUsd: 30_000,
      costHighUsd: 120_000,
      liftPoints: 45,
    });
  }
  if (!improvements.septic && !features.has('septic')) {
    gaps.push({
      label: 'Conventional septic system (1k gal tank + leach field)',
      costLowUsd: 6_000,
      costHighUsd: 14_000,
      liftPoints: 8,
    });
  }
  if (!improvements.electric && !features.has('electric')) {
    gaps.push({
      label: 'Run grid power (per pole, varies by distance)',
      costLowUsd: 4_000,
      costHighUsd: 30_000,
      liftPoints: 6,
    });
  }

  const verdict =
    score >= 80 ? 'Move-in ready or close — minimal extra build needed.' :
    score >= 60 ? 'Outbuilding starting point + utilities; cabin/home is the next step.' :
    score >= 40 ? 'Bare land with utilities; substantial build ahead.' :
    'Raw land — every system from scratch.';

  return { key: 'shelter', label: 'Shelter', score, weight: AXIS_WEIGHTS.shelter, verdict, gaps };
};

const resilienceAxis = (p: Property): Axis => {
  let score = 75; // baseline assumes "no major flags"
  const flood = p.geoEnrichment?.flood;
  const features = new Set(p.features ?? []);

  if (flood?.isSFHA) score -= 30; // FEMA SFHA = 100-yr floodplain
  if (flood?.floodZone === 'X' || flood?.floodZone === 'X500') score += 5;

  // Drought proxy: arid SW interior
  const lat = p.location?.lat ?? 0;
  const lng = Math.abs(p.location?.lng ?? 0);
  if (lat > 30 && lat < 38 && lng > 103 && lng < 116) score -= 12;

  // Slope = wildfire + erosion proxy
  const slope = p.geoEnrichment?.soil?.slopePercent ?? 0;
  if (slope > 20) score -= 10;
  else if (slope > 12) score -= 5;

  // Owner type: ag exemption proxy via acreage. TX needs ≥10ac for ag-use.
  const acres = p.acreage ?? 0;
  if (acres >= 10) score += 8; // likely ag-exempt eligible
  if (features.has('no_hoa')) score += 5;

  score = Math.max(0, Math.min(100, Math.round(score)));

  const gaps: Gap[] = [];
  if (acres >= 10) {
    gaps.push({
      label: 'File for ag exemption (TX: ~80% property-tax reduction)',
      costLowUsd: 0,
      costHighUsd: 500,
      liftPoints: 8,
    });
  }
  if (slope > 12) {
    gaps.push({
      label: 'Defensible-space clearing + fire breaks',
      costLowUsd: 2_000,
      costHighUsd: 10_000,
      liftPoints: 6,
    });
  }

  const verdict =
    score >= 80 ? 'Few resilience flags — solid foundation for off-grid living.' :
    score >= 60 ? 'Manageable risks. Ag exemption + defensible space close most gaps.' :
    score >= 40 ? 'Notable flags — flood, fire, drought, or zoning limit autonomy.' :
    'High-risk parcel. Self-sufficiency fights against the land here.';

  return { key: 'resilience', label: 'Resilience', score, weight: AXIS_WEIGHTS.resilience, verdict, gaps };
};

// ── Composite ─────────────────────────────────────────────────────────

export const computeSelfSufficiency = (p: Property): SelfSufficiencyReport => {
  const axes = [foodAxis(p), waterAxis(p), energyAxis(p), shelterAxis(p), resilienceAxis(p)];
  const composite = Math.round(
    axes.reduce((sum, a) => sum + a.score * a.weight, 0),
  );
  const weakest = [...axes].sort((a, b) => a.score - b.score)[0];

  // Aggregate cost-to-full from gaps. Cap per-axis lift at gap-to-85.
  let costLow = 0;
  let costHigh = 0;
  const projectedAxes = axes.map((axis) => {
    let projected = axis.score;
    for (const gap of axis.gaps) {
      if (projected >= 85) break;
      projected = Math.min(100, projected + gap.liftPoints);
      costLow += gap.costLowUsd;
      costHigh += gap.costHighUsd;
    }
    return { ...axis, projected };
  });
  const potentialComposite = Math.round(
    projectedAxes.reduce((sum, a) => sum + a.projected * a.weight, 0),
  );

  return {
    composite,
    axes,
    weakest,
    costToFullLowUsd: costLow,
    costToFullHighUsd: costHigh,
    potentialComposite,
  };
};
